examples/Fine_tuning_for_function_calling.ipynb

```diff
@@ -0,0 +1,1248 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Fine tuning for function calling (with synthetic data)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook covers how to finetune for function calling to increase function calling accuracy and reliability.\\\n",
+    "For reference, the cookbook recipe for function calling is here: https://github.com/openai/openai-cookbook/blob/main/examples/How_to_call_functions_with_chat_models.ipynb \\\n",
+    "and a cookbook recipe for fine tuning chat responses is here: https://github.com/openai/openai-cookbook/blob/main/examples/How_to_finetune_chat_models.ipynb"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For context, from the function calling notebook above:\n",
+    "> `functions` is an optional parameter in the Chat Completion API which can be used to provide function specifications. The purpose of this is to enable models to generate function arguments which adhere to the provided specifications. Note that the API will not actually execute any function calls. It is up to developers to execute function calls using model outputs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Overview\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook contains three sections\n",
+    "- **Assessing baseline function calling performance:** Evaluating an out-of-the-box 3.5 Turbo model on our given function (let's assume that for latency + cost reasons we cannot use GPT-4 for a drone copilot)\n",
+    "- **Generating synthetic data:** Using GPT-4 to create 'golden' set of prompts and function invocations to use as training data\n",
+    "- **Fine-tuning**: Running the fine tuning job, and evaluating the fine-tuned model\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note: *This notebook provides an example of how to create synthetic training data for fine tuning for function calling given just a list of functions. While real-world production test evals are preferable, this method produces strong results and can (and should) be used in conjuction with real-world training data.*"
```

Comment:
(and should) <- this might vary, could be cases it doesn't help, so we might wanna leave it at just "can"